[Feature] Support IPU

HuDi2018 opened the pull request:
ipu model training and inference are basing on static graph.
so we need to run model training/inference once to compile graph on IPU.
all the convert and config parsing things are controlled by ipu_runner

## Motivation

implement mmcv on IPU and support a demo(ViT training) on IPU

## Modification

add an ipu_runner and some help functions and classes in runner/ipu

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

use case is in  a PR of mmclassification which will be referenced in comments

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

wHao-Wu committed:
implement MMCV on IPU

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/open-mmlab/mmcv?pullRequest=1780) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/open-mmlab/mmcv?pullRequest=1780) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: HuDi2018<br/>:white_check_mark: wHao-Wu<br/>:x: root<br/><hr/>**root** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/open-mmlab/mmcv?pullRequest=1780) it.</sub>

zhouzaida commented on mmcv/runner/__init__.py:
```diff
@@ -62,5 +63,5 @@
     'ModuleDict', 'ModuleList', 'GradientCumulativeOptimizerHook',
     'GradientCumulativeFp16OptimizerHook', 'DefaultRunnerConstructor',
-    'SegmindLoggerHook'
+    'SegmindLoggerHook', 'IpuEpochBasedRunner', 'IpuIterBasedRunner', 'ipu'
```
```suggestion
    'SegmindLoggerHook', 'IPUEpochBasedRunner', 'IPUIterBasedRunner', 'ipu'
```

zhouzaida replied:
Should we import 'ipu' first and then add it to `__all__`?

HumberMe replied:
sorry, ipu is a redundant import, I will del it

zhouzaida commented on mmcv/runner/__init__.py:
```diff
@@ -37,6 +37,7 @@
                         build_optimizer_constructor)
 from .priority import Priority, get_priority
 from .utils import get_host_info, get_time_str, obj_from_dict, set_random_seed
+from .ipu_runner import IpuEpochBasedRunner, IpuIterBasedRunner
```
```suggestion
from .ipu_runner import IPUEpochBasedRunner, IPUIterBasedRunner
```

zhouzaida commented on mmcv/runner/fp16_utils.py:
```diff
@@ -102,9 +102,10 @@ def auto_fp16_wrapper(old_func):
+            if not isinstance(args[0], _supported_types):
                 raise TypeError('@auto_fp16 can only be used to decorate the '
-                                'method of nn.Module')
+                                'method of nn.Module andy types you specified')
```
```suggestion
                                'method of nn.Module and types you specified')
```

zhouzaida commented on mmcv/runner/ipu/__init__.py:
```diff
@@ -0,0 +1,20 @@
+except ImportError:
+    IPU_MODE = False
+
+if IPU_MODE:
```
Suggest moving this file to `mmcv/utils` and renaming it to `ipu_wrapper.py`

zhouzaida commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,92 @@
+                self.detect_anomalous_parameters(
+                    runner.outputs['loss'], runner)
+            if self.grad_clip is not None:
+                raise NotImplementedError('IPU not supports gradient clip now')
```
```suggestion
                raise NotImplementedError('IPU does not support gradient clip')
```

zhouzaida commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,92 @@
+if (TORCH_VERSION != 'parrots'
+        and digit_version(TORCH_VERSION) >= digit_version('1.6.0')):
+    @HOOKS.register_module()
+    class IpuFp16OptimizerHook(OptimizerHook):
```
```suggestion
    class IPUFp16OptimizerHook(OptimizerHook):
```

HuDi2018 committed:
change Ipu to IPU, move IPU_MODE to utils/ipu_wrapper

HuDi2018 committed:
change Ipu to IPU, move IPU_MODE to utils/ipu_wrapper

HuDi2018 committed:
rm tmp_test.sh

HuDi2018 committed:
fix some lint errors and warnings

HuDi2018 committed:
remove diff.txt and fix a typo

HuDi2018 committed:
merge to remote branch

HuDi2018 committed:
Merge branch 'master' into submit1

HuDi2018 committed:
fix copyright, spelling mistake, double quote, blank space and some problem in test_runner.py

HuDi2018 committed:
Merge branch 'submit1' into submit

HuDi2018 committed:
del ipu runner test on cpu and gpu mode

zhouzaida commented on mmcv/runner/ipu/__init__.py:
```diff
@@ -0,0 +1,18 @@
+
+
+if IPU_MODE:
+    from .util import parse_ipu_options, ipu_model_wrapper,\
+        build_from_cfg_with_wrapper, model_sharding,\
+        recomputation_checkpoint
```
```suggestion
    from .util import (parse_ipu_options, ipu_model_wrapper,
        build_from_cfg_with_wrapper, model_sharding,
        recomputation_checkpoint)
```

zhouzaida commented on mmcv/runner/ipu/__init__.py:
```diff
@@ -0,0 +1,18 @@
+    from .util import parse_ipu_options, ipu_model_wrapper,\
+        build_from_cfg_with_wrapper, model_sharding,\
+        recomputation_checkpoint
+    from .hooks import wrap_optimizer_hook, IPUFp16OptimizerHook,\
+        wrap_lr_update_hook
```
```suggestion
    from .hooks import (wrap_optimizer_hook, IPUFp16OptimizerHook,
        wrap_lr_update_hook)
```

zhouzaida commented on mmcv/runner/ipu/dataloder.py:
```diff
@@ -0,0 +1,83 @@
+                 mode=poptorch.DataLoaderMode.AsyncRebatched,
+                 rebatched_worker_size=128,
+                 **kwargs):
+        # lazy init
```
Suggest adding a more detailed comment to explain why we need to initialize lazily.

zhouzaida commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,93 @@
+from mmcv.runner.hooks import HOOKS, OptimizerHook, LrUpdaterHook
+
+
+def wrap_lr_update_hook(lr_hook_class,):
```
```suggestion
def wrap_lr_update_hook(lr_hook_class):
```

zhouzaida replied:
Suggest adding a summary or description to explain why we need to wrap `lr_update_hook`, which will be helpful to users or other developers.

zhouzaida commented on mmcv/runner/fp16_utils.py:
```diff
@@ -63,7 +63,7 @@ def cast_tensor_type(inputs, src_type, dst_type):
 
 
-def auto_fp16(apply_to=None, out_fp32=False):
+def auto_fp16(apply_to=None, out_fp32=False, supported_types=[]):
```
Add `supported_types` to line 79 and append `New in version 1.4.8.` in the end.

https://github.com/open-mmlab/mmcv/blob/6289b6f922f3b251791ef7cb5fd05d0c900b4750/mmcv/runner/hooks/logger/text.py#L52

zhouzaida commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,93 @@
+
+
+if (TORCH_VERSION != 'parrots'
+        and digit_version(TORCH_VERSION) >= digit_version('1.6.0')):
```
Could `IPUFp16OptimizerHook` be supported in torch1.6.0?

zhouzaida commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,93 @@
+                     loss_scale=512.,
+                     distributed=True):
+            assert grad_clip is None,\
+                'IPU mode not support grad_clip currently'
```
```suggestion
                'IPU mode does not support `grad_clip` currently'
```

zhouzaida commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,93 @@
+            assert coalesce,\
+                'implemented all reduce in distributed training currently'
+            assert bucket_size_mb == -1,\
+                'no bucket_size_mb can be set in IPU mode'
```
```suggestion
                '`bucket_size_mb` should not be set in IPU mode'
```

zhouzaida commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,93 @@
+            self._scale_update_param = None
+            if loss_scale == 'dynamic':
+                raise NotImplementedError(
+                    'IPU mode not support dynamic loss scale currently')
```
```suggestion
                    'IPU mode does not support dynamic loss scale currently')
```

zhouzaida commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,93 @@
+        def after_train_iter(self, runner):
+            pass
+
+else:
+    raise RuntimeError('The IPU mode only supports torch1.10 and above')
```
It seems like the `else` clause here is not correct.

zhouzaida commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,93 @@
+    return ipu_lr_hook_class
+
+
+def wrap_optimizer_hook(optimizer_hook_class,):
```
Same as above.

zhouzaida commented on mmcv/utils/__init__.py:
```diff
@@ -17,6 +17,7 @@
                       check_python_script)
 from .timer import Timer, TimerError, check_time
 from .version_utils import digit_version, get_git_hash
+from .ipu_wrapper import IPU_MODE
```
suggest renaming `IPU_MODE` to `is_ipu_pytorch` which is consistent with `is_rocm_pytorch`.

HuDi2018 committed:
Update mmcv/runner/ipu/__init__.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/__init__.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/hooks.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/hooks.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/hooks.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/hooks.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/hooks.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/hooks.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
add docstrings to wrap_lr_update_hook, wrap_optimizer_hook, IPUDataloader and IPUBaseRunner
add a detailed explaination for lazy init of IPUDataloder
refine the docstrings of function auto_fp16
change the compatible version from pytorch 1.10.0 to 1.6.0

HuDi2018 committed:
add  getting start doc

HuDi2018 committed:
make ipu_runner support poptorch.dataloader and pytorch.dataloder

HuDi2018 committed:
add an assertion for eval_model partition

zhouzaida commented on mmcv/runner/ipu/dataloder.py:
```diff
@@ -0,0 +1,134 @@
+    """
+
+    if not isinstance(batch, Sequence):
+        raise TypeError(f'{batch.dtype} is not supported.')
```
```suggestion
        raise TypeError(f'`batch` should be a sequnece, but got {type(bajtch)}.')
```

HuDi2018 committed:
updata a typeerror in ipu.dataloder

zhouzaida commented on mmcv/runner/ipu/dataloder.py:
```diff
@@ -0,0 +1,134 @@
+    """Puts each data field into a tensor/DataContainer with outer dimension
+    batch size.
+
+    Extend default_collate to add support for
+    :type:`~mmcv.parallel.DataContainer`. There are 3 cases.
```
If `collate` does not support `DataContainer`, those lines can be removed.

zhouzaida commented on mmcv/runner/ipu/dataloder.py:
```diff
@@ -0,0 +1,134 @@
+        raise TypeError('DataContainer is not supported in ipu data loader.')
+    elif isinstance(batch[0], Sequence):
+        transposed = zip(*batch)
+        _list = []
+        for samples in transposed:
+            if not isinstance(samples[0], DataContainer):
+                _list.append(collate(samples, samples_per_gpu))
+        return _list
```
```suggestion
        collated_batch = []
        for samples in transposed:
            assert not isinstance(samples[0], DataContainer)
            collated_batch.append(collate(samples, samples_per_gpu))
        return collated_batch 
```

zhouzaida commented on mmcv/runner/ipu/dataloder.py:
```diff
@@ -0,0 +1,134 @@
+    if not isinstance(batch, Sequence):
+        raise TypeError(f'{batch.dtype} is not supported.')
+
+    if isinstance(batch[0], DataContainer):
+        raise TypeError('DataContainer is not supported in ipu data loader.')
```
```suggestion
    if isinstance(batch[0], DataContainer):
        # `DataContainer` will be supported in the future.
        raise TypeError('DataContainer is not supported in ipu data loader.')
```

HuDi2018 committed:
made some changes to ipu.dataloader

zhouzaida commented on mmcv/runner/ipu/dataloder.py:
```diff
@@ -0,0 +1,134 @@
+
+
+class IPUDataloader(poptorch.DataLoader):
+    """ Thin wrapper around the traditional `torch.utils.data.DataLoader`
+    to abstract away some of the batch sizes calculationsand speed up
+    data loading.
```
We need to refine the summary.

zhouzaida commented on mmcv/runner/ipu/dataloder.py:
```diff
@@ -0,0 +1,134 @@
+
+
+class IPUDataloader(poptorch.DataLoader):
+    """ Thin wrapper around the traditional `torch.utils.data.DataLoader`
+    to abstract away some of the batch sizes calculationsand speed up
+    data loading.
```
```suggestion
    """Thin wrapper around the traditional `torch.utils.data.DataLoader`
    to abstract away some of the batch sizes calculationsand speed up
    data loading.
```

zhouzaida replied:
What is the meaning of `calculationsand`?

zhouzaida commented on mmcv/runner/ipu/dataloder.py:
```diff
@@ -0,0 +1,134 @@
+    to abstract away some of the batch sizes calculationsand speed up
+    data loading.
+
+    If this data loader is used in a distributed execution environment, it will
+    ensure that each process uses a different subset of the dataset, providing
+    you first call ``options.randomSeed(N)`` with an integer N which is the
+    same across all hosts.
+    Args:
```
```suggestion
    If this data loader is used in a distributed execution environment, it will
    ensure that each process uses a different subset of the dataset, providing
    you first call ``options.randomSeed(N)`` with an integer N which is the
    same across all hosts.

    Args:
```

HuDi2018 committed:
fix some format problems

zhouzaida commented on mmcv/runner/ipu/dataloder.py:
```diff
@@ -0,0 +1,138 @@
+                collated_batch.append(collate(samples, samples_per_gpu))
+        return collated_batch
+    elif isinstance(batch[0], Mapping):
+        collated_dic = {}
+        for key in batch[0]:
+            if not isinstance(batch[0][key], DataContainer):
+                collated_dic[key] = collate([d[key] for d in batch])
+        return collated_dic
```
```suggestion
        collated_batch = {}
        for key in batch[0]:
            if not isinstance(batch[0][key], DataContainer):
                collated_batch[key] = collate([d[key] for d in batch])
        return collated_batch
```

HuDi2018 committed:
add some notes for datacontainer in ipu.dataloder

HuDi2018 committed:
refine kwargs of dataloder

hellock commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,134 @@
+            model,
+            ipu_options={},
+            modules_to_record=[],
+            ipu_model_cfg={},
```
Using empty list or dict as default values is unsafe.

HumberMe replied:
Is it because of recursion or multithreading？

hellock commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,262 @@
+            opts_node(_cfg)
+        else:
+            error_msg = 'opts_node type {} not supported'.format(
+                type(opts_node))
```
f-string is preferred in OpenMMLab projects

hellock commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,262 @@
+        raise NotImplementedError(error_msg)
+
+
+def parse_ipu_options(ipu_options):
```
Docstring is missing.

hellock commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,262 @@
+                    Note that the ``ipu_id`` is an index
+                    in a multi-IPU device within PopTorch, and is
+                    separate and distinct from the device ids used by
+                    ``gc-info``.
```
The docstring and argument list do not match.

hellock commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,262 @@
+        else:
+            return poptorch.recomputationCheckpoint(outputs)
+
+    for idx, (_name, _module) in enumerate(model.named_modules()):
```
`idx` is unused, thus there is no need to use `enumerate()`.

hellock commented on mmcv/runner/ipu/dataloder.py:
```diff
@@ -0,0 +1,150 @@
+    If this data loader is used in a distributed execution environment, it will
+    ensure that each process uses a different subset of the dataset, providing
+    you first call ``options.randomSeed(N)`` with an integer N which is the
+    same across all hosts.
```
Better to describe the brief differences from PyTorch DataLoader in the docstring.

hellock commented on mmcv/runner/ipu/dataloder.py:
```diff
@@ -0,0 +1,150 @@
+        """
+        # lazy init: sometimes, we cannot get IPU options when make data
+        #            loader
+        if mode == 'async':
```
Better to validate the values of `mode` because there is no fallback or else branch after Line146.

hellock commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,104 @@
+    IPU's clip-norm implementation is different from pytorch, so there
+    should be an error raised when using clip-norm.
+    """
+    assert optimizer_hook_class == OptimizerHook,\
```
issubclass instead of ==?

HumberMe replied:
only support OptimizerHook, currently not support subclass of OptimizerHook like GradientCumulativeOptimizerHook

hellock commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,104 @@
+from mmcv.runner.hooks import HOOKS, OptimizerHook, LrUpdaterHook
+
+
+def wrap_lr_update_hook(lr_hook_class):
```
Use `functools.wraps` in decorators.

HumberMe replied:
currently it is only used to wrap a function, no need to use as a decorator

hellock commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,713 @@
+        self._warned_not_contiguous_input = False
+
+
+class ComplexDataManager:
```
The docstring is missing.

HuDi2018 committed:
add some docstrings and refine some docstrings
replace format with  f-string
remove mode process in __init__ of dataloder
no more use {} and [] as default value of function arguments

HuDi2018 committed:
change double quote to single quote

zhouzaida commented on mmcv/runner/fp16_utils.py:
```diff
@@ -63,7 +63,7 @@ def cast_tensor_type(inputs, src_type, dst_type):
 
 
-def auto_fp16(apply_to=None, out_fp32=False):
+def auto_fp16(apply_to=None, out_fp32=False, supported_types=[]):
```
```suggestion
def auto_fp16(apply_to=None, out_fp32=False, supported_types=None):
```

zhouzaida replied:
It is a dangerous way to use mutable `[]` as a default value.

zhouzaida commented on mmcv/runner/fp16_utils.py:
```diff
@@ -102,9 +104,10 @@ def auto_fp16_wrapper(old_func):
             # check if the module has set the attribute `fp16_enabled`, if not,
             # just fallback to the original method.
-            if not isinstance(args[0], torch.nn.Module):
+            _supported_types = tuple([torch.nn.Module] + supported_types)
+            if not isinstance(args[0], _supported_types):
```
```suggestion
            if not isinstance(args[0], supported_types):
```

zhouzaida commented on mmcv/runner/fp16_utils.py:
```diff
@@ -102,9 +104,10 @@ def auto_fp16_wrapper(old_func):
+            if not isinstance(args[0], _supported_types):
                 raise TypeError('@auto_fp16 can only be used to decorate the '
-                                'method of nn.Module')
+                                'method of nn.Module and types you specified')
```
```suggestion
                                f'method of those classes {supported_types}')
```

zhouzaida commented on mmcv/runner/fp16_utils.py:
```diff
@@ -76,7 +76,9 @@ def auto_fp16(apply_to=None, out_fp32=False):
             `None` indicates all arguments.
         out_fp32 (bool): Whether to convert the output back to fp32.
-
+        supported_types (list[Class], optional): class types to support besides
+            torch.nn.Module
+            `New in version 1.4.8.`
```
```suggestion
        supported_types (tuple): Classes can be decorated by ``auto_fp16``.
            `New in version 1.5.0.`
```

zhouzaida commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,104 @@
+
+
+def wrap_lr_update_hook(lr_hook_class):
+    """A wrapper function to wrap any subclass of LrUpdaterHook
+    IPU needs extra operations to upload optimizer setting,
+    this wrapper will override function(_set_lr) of a subclass of
+    LrUpdaterHook.
```
```suggestion
    """A wrapper function to wrap any subclass of LrUpdaterHook.
    
    IPU needs extra operations to upload optimizer settings,
    this wrapper will override function(_set_lr) of a subclass of
    LrUpdaterHook.
```

zhouzaida commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,104 @@
+
+    class ipu_lr_hook_class(lr_hook_class):
+        def _set_lr(self, runner, *args, **kwargs):
+            result = super()._set_lr(runner, *args, **kwargs)
+            assert result is None  # _set_lr should return nothing
```
Why do we need to assert that the result is None? Should we directly call `super()._set_lr(runner, *args, **kwargs)` and return nothing?

zhouzaida commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,104 @@
+
+
+def wrap_optimizer_hook(optimizer_hook_class,):
+    """A wrapper function to wrap OptimizerHook
+    This is an non-intrusive implementation of wrapping optimizer hook
+    (or you need to change every config file to use IPU optimizer hook)
+    IPU's clip-norm implementation is different from pytorch, so there
+    should be an error raised when using clip-norm.
+    """
```
```suggestion
    """A wrapper function to wrap OptimizerHook.
    
    This is a non-intrusive implementation of wrapping optimizer hook
    (or you need to change every config file to use IPU optimizer hook)
    IPU's clip-norm implementation is different from pytorch, so there
    should be an error raised when using clip-norm.
    """
```

zhouzaida commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,104 @@
+    IPU's clip-norm implementation is different from pytorch, so there
+    should be an error raised when using clip-norm.
+    """
+    assert optimizer_hook_class == OptimizerHook,\
+        f'OptimizerHook type used is:{str(optimizer_hook_class)},\
+             not supported now'
```
```suggestion
    if optimizer_hook_class is not OptimizerHook:
        raise ValueError(
            f'Only OptimizerHook is supported , bug got {optimizer_hook_class}'
```

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,136 @@
+            raise TypeError(
+                'if you want to implement data parallelism '
+                'at the module level on IPU, '
+                'use IPU option: replicationFactor')
+        ipu_options = ipu_options or {}
```
```suggestion
                'use IPU option: replicationFactor')
                
        ipu_options = ipu_options or {}
```

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,136 @@
+
+
+class IPUBaseRunner(metaclass=ABCMeta):
+    """A base runner for IPU device, should be inherited with
+    base_runner.BaseRunner
+    This runner has some extra processes for IPU which are showed below:
+    1. Parse options for IPU
+    2. wrap pytorch model for IPU
+    3. Raise errors while encountering illegal ussage
+    4. Input IPU options and initialize dataloader if finding a instance of
+       of IPUDataloader
```
```suggestion
    """All subclasses run on IPU device should inherit this base class and 
    ``base_runner.BaseRunner``.
    
    This runner has some extra processes for IPU which are shown below:
    
    1. Parse options for IPU
    2. wrap pytorch model for IPU
    3. Raise errors while encountering illegal usage
    4. Input IPU options and initialize dataloader if finding an instance of
       of IPUDataloader
```

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,136 @@
+        # process options of ipu
+        if IPU_MODE:
+            self.ipu_options = parse_ipu_options(ipu_options)
+            # self.data_loader = wrap_data_loader(self.data_loader)
```
This line should be removed if we do not wrap it.

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,136 @@
+
+        super().run(data_loaders, workflow, *args, **kwargs)
+
+    def get_ipu_opts(self, mode):
```
If mode is neither `train` nor `val`, should an exception be raised?

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,136 @@
+    def run(self, data_loaders, workflow, *args, **kwargs):
+        for i, flow in enumerate(workflow):
+            mode, iters = flow
+            # initialize IPU dataloder if not initialized
+            data_loaders[i].init(options=self.get_ipu_opts(mode))
```
Should we assert the data_loader must be `IPUDataloader`?

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,136 @@
+
+@RUNNERS.register_module()
+class IPUEpochBasedRunner(IPUBaseRunner, EpochBasedRunner):
+    """Epoch-based Runner for IPU.
+    The Inheritance order(MRO) is:
+    IPUEpochBasedRunner -> IPUBaseRunner -> EpochBasedRunner -> BaseRunner
+    This runner train models epoch by epoch.
+    """
+    pass
```
```suggestion
    """Epoch-based Runner for IPU.
    
    The Inheritance order(MRO) is:
    IPUEpochBasedRunner -> IPUBaseRunner -> EpochBasedRunner -> BaseRunner
    This runner train models epoch by epoch.
    """
    pass
```

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,136 @@
+
+@RUNNERS.register_module()
+class IPUIterBasedRunner(IPUBaseRunner, IterBasedRunner):
+    """Iteration-based Runner for IPU.
+    The Inheritance order(MRO) is:
+    IPUIterBasedRunner -> IPUBaseRunner -> IterBasedRunner -> BaseRunner
+    This runner train models iteration by iteration.
+    """
```
```suggestion
    """Iteration-based Runner for IPU.
    
    The Inheritance order(MRO) is:
    IPUIterBasedRunner -> IPUBaseRunner -> IterBasedRunner -> BaseRunner
    This runner train models iteration by iteration.
    """
```

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,136 @@
+
+
+@RUNNERS.register_module()
+class IPUIterBasedRunner(IPUBaseRunner, IterBasedRunner):
```
Can we simplify this inheritance with `single inheritance`?

HumberMe replied:
no idea how to simplify.
because we need inheritance from IterBasedRunner and IPUBaseRunner.
or we need to implement same functions twice in IPUIterBasedRunner and IPUEpochBasedRunner.

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,265 @@
+        obj_cls = obj_type
+    else:
+        raise TypeError(
+            f'type must be a str or valid type, but got {type(obj_type)}')
+    wrapped_obj_cls = obj_cls if wrapper_func is None \
+        else wrapper_func(obj_cls)
```
```suggestion
            f'type must be a str or valid type, but got {type(obj_type)}')
            
    if wrapper_func is None:
        wrapped_obj_cls = obj_cls
    else:
        wrapped_obj_cls = wrapper_func(obj_cls)
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,265 @@
+            of split edge
+
+    Returns:
+        model (pytorch.nn.Module): split model
```
```suggestion
        nn.Module: Split model.
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,265 @@
+    """split models in-place into multi-IPUs
+
+    Args:
+        model (pytorch.nn.Module): the target model to be split
+        split_edges (dict): model layer names or layer numbers
+            of split edge
```
```suggestion
        model (nn.Module): The target model to be split.
        split_edges (dict): Model layer names or layer numbers
            of split edge.
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,265 @@
+
+
+def model_sharding(model, split_edges):
+    """split models in-place into multi-IPUs
```
```suggestion
    """split models in-place into multi-IPUs.
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,265 @@
+    if len(split_edges) == 0:
+        return model
+    assert isinstance(split_edges, list)
+    spilt_edges_dic = {ele['layer_to_call']: ele for ele in split_edges}
```
```suggestion
    spilt_edges_dict = {edge['layer_to_call']: edge for edge in split_edges}

```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,265 @@
+    return model
+
+
+def model_sharding(model, split_edges):
```
An example should be added to show the format of split_edges.

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,265 @@
+        if edge is not None:
+            poptorch.BeginBlock(_module, edge.get(
+                'user_id', _name), edge['ipu_id'])
+    # check all split_edges are used
+    split_edge_names = list(spilt_edges_dic.keys())
+    assert len(spilt_edges_dic) == 0,\
+        f'split_edges: {split_edge_names} are not contained in the model'
```
```suggestion

    # ensure all split_edges are used
    if len(split_edges_dic) > 0:
        raise RuntimeError(
            f'split_edges: {list(spilt_edges_dic.keys())} are not contained in the model')
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,265 @@
+        return model
+    assert isinstance(split_edges, list)
+    spilt_edges_dic = {ele['layer_to_call']: ele for ele in split_edges}
+    for idx, (_name, _module) in enumerate(model.named_modules()):
+        assert not (idx in spilt_edges_dic and _name in spilt_edges_dic),\
+            f'The same layer is referenced twice while doing model partition: \
+                idx is {idx} and name is {_name}'
+        edge = spilt_edges_dic.pop(_name, None)
+        edge = spilt_edges_dic.pop(idx, edge)
+        if edge is not None:
+            poptorch.BeginBlock(_module, edge.get(
+                'user_id', _name), edge['ipu_id'])
```
```suggestion
    for idx, (name, module) in enumerate(model.named_modules()):
        if idx in spilt_edges_dic and name in spilt_edges_dic:
            raise ValueError(
                f'The same layer is referenced twice while doing model partition: '
                'idx is {idx} and name is {name}'
            
        edge = spilt_edges_dic.pop(name, None)
        edge = spilt_edges_dic.pop(idx, edge)
        if edge is not None:
            poptorch.BeginBlock(module, edge.get(
                'user_id', name), edge['ipu_id'])
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,265 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import copy
+import inspect
+import torch
```
```suggestion
import torch
import torch.nn as nn
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,265 @@
+import inspect
+import torch
+import popart
+import poptorch
+from mmcv.utils import Registry
```
```suggestion
import poptorch

from mmcv.utils import Registry
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,265 @@
+    return model
+
+
+def recomputation_checkpoint(model: torch.nn.Module, module_names=[])\
```
```suggestion
def recomputation_checkpoint(model: nn.Module, module_names=[])\
```

zhouzaida replied:
It is not recommended to use mutable `[]` as a default value.

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,265 @@
+
+def recomputation_checkpoint(model: torch.nn.Module, module_names=[])\
+     -> torch.utils.hooks.RemovableHandle:
+    """Annotates the output of a module to be checkpointed instead of
+        recomputed"""
```
```suggestion
    """Annotates the output of a module to be checkpointed instead of
        recomputed.
        
    Args:
        model (nn.Module): ...
        module_names (list): ...
    """
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,265 @@
+        else:
+            return poptorch.recomputationCheckpoint(outputs)
+
+    for _name, _module in model.named_modules():
+        if _name in module_names:
+            _module.register_forward_hook(recompute_outputs)
+            module_names.remove(_name)
```
```suggestion
    for name, module in model.named_modules():
        if name in module_names:
            module.register_forward_hook(recompute_outputs)
            module_names.remove(name)
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,265 @@
+        ipu_model_cfg={},
+        fp16_cfg=None
+        ):
+    # TrainEvalModel will shallow copy the model,
```
docstring is missing.

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,265 @@
+        if optimizer is not None:
+            optimizer.loss_scaling = loss_scale
+            if fp16_cfg.get('velocity_accum_type', False):
+                optimizer.velocity_accum_type =\
+                    torch.half if fp16_cfg['velocity_accum_type'] == 'half'\
+                    else torch.float32
```
```suggestion
                if fp16_cfg['velocity_accum_type'] == 'half':
                    optimizer.velocity_accum_type = torch.half
                else:
                    optimizer.velocity_accum_type = torch.float32
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,265 @@
+                    torch.half if fp16_cfg['velocity_accum_type'] == 'half'\
+                    else torch.float32
+            if fp16_cfg.get('accum_type', False):
+                optimizer.accum_type =\
+                    torch.half if fp16_cfg['accum_type'] == 'half'\
+                    else torch.float32
```
```suggestion
                if fp16_cfg['accum_type'] == 'half':
                    optimizer.accum_type = torch.half
                else:
                    optimizer.accum_type = torch.float32
```

HuDi2018 committed:
Update mmcv/runner/fp16_utils.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/fp16_utils.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/fp16_utils.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/fp16_utils.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/hooks.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu_runner.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/util.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/util.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/util.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/util.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/util.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
code review 1

HuDi2018 committed:
code review 2

zhouzaida commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,105 @@
+def wrap_lr_update_hook(lr_hook_class):
+    """A wrapper function to wrap any subclass of LrUpdaterHook.
+    
+    IPU needs extra operations to upload optimizer settings,
+    this wrapper will override function(_set_lr) of a subclass of
```
```suggestion
    IPU needs extra operations to upload optimizer settings.
    This wrapper will override function(_set_lr) of a subclass of
```

zhouzaida commented on mmcv/runner/ipu/hooks.py:
```diff
@@ -0,0 +1,105 @@
+    class ipu_lr_hook_class(lr_hook_class):
+        def _set_lr(self, runner, *args, **kwargs):
+            super()._set_lr(runner, *args, **kwargs)
+            runner.model.setOptimizer(runner.optimizer)
```
```suggestion
            # apply the new optimizer values to the model
            runner.model.setOptimizer(runner.optimizer)
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,316 @@
+
+
+def _opts_assigner(_cfg, opts_node):
+    # set popart.options by config
+    # _cfg: dict, python data type
+    # opts_node: python module or function
```
```suggestion
    """Set popart.options by config.
    
    Args:
        cfg (dict): xxx.
        opts_node (callable): xxx.
    """
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,316 @@
+        raise type(e)(f'{wrapped_obj_cls.__name__}: {e}')
+
+
+def _opts_assigner(_cfg, opts_node):
```
```suggestion
def _options_assigner(cfg, options_node):
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,316 @@
+    # set popart.options by config
+    # _cfg: dict, python data type
+    # opts_node: python module or function
+    if isinstance(_cfg, dict):
+        for _key in _cfg:
+            _opts_assigner(_cfg[_key], getattr(opts_node, _key))
+    elif isinstance(_cfg, (int, float, str, list)):
+        if callable(opts_node):
+            opts_node(_cfg)
```
```suggestion
    if isinstance(cfg, dict):
        for key, value in cfg.items():
            opts_assigner(value, getattr(options_node, key))
    elif isinstance(cfg, (int, float, str, list)):
        if callable(options_node):
            options_node(cfg)
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,316 @@
+        split_edges (list): Model layer names or layer numbers
+            of split edge.
+
+        Examples:
+            >>> split_edges = [
+            ...     # layer_to_call is name of layer and ipu_id
+            ...     # if the id of ipu to map
+            ...     dict(layer_to_call='model.conv1', ipu_id=0),
+            ...     dict(layer_to_call='model.conv3', ipu_id=1)]
+            >>> sharding_model = model_sharding(torch_model, split_edges)
```
```suggestion
    Examples:
        >>> split_edges = [
        ...     # layer_to_call is name of layer and ipu_id
        ...     # if the id of ipu to map
        ...     dict(layer_to_call='model.conv1', ipu_id=0),
        ...     dict(layer_to_call='model.conv3', ipu_id=1)]
        >>> sharding_model = model_sharding(torch_model, split_edges)
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,316 @@
+
+    Args:
+        model (nn.Module): The target model to be split.
+        split_edges (list): Model layer names or layer numbers
+            of split edge.
```
```suggestion
        split_edges (list of dict): Model layer names or layer numbers
            of split edge.
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,316 @@
+    return model
+
+
+def recomputation_checkpoint(model: nn.Module, module_names=[])\
```
```suggestion
def recomputation_checkpoint(model: nn.Module, module_names)\
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,316 @@
+    """Annotates the output of a module to be checkpointed instead of
+        recomputed
+
+    If recompute mode is enabled, ipu will release the activations of
```
```suggestion
    If recomputation mode is enabled, ipu will release the activations of
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,316 @@
+
+    # check all module_names are used
+    assert len(module_names) == 0,\
+        f'split_edges: {module_names} are not contained in the model'
```
```suggestion
        f'module_names: {module_names} are not contained in the model'
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,316 @@
+
+def recomputation_checkpoint(model: nn.Module, module_names=[])\
+     -> torch.utils.hooks.RemovableHandle:
+    """Annotates the output of a module to be checkpointed instead of
+        recomputed
```
```suggestion
    """Annotates the output of a module to be checkpointed instead of
    recomputed.
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,316 @@
+    Args:
+        model (nn.Module): The target model to be converted.
+        opts (dict): IPU options.
+        optimizer (torch.optim | optional): torch optimizer, necessary
```
```suggestion
        optimizer (torch.optim, optional): torch optimizer, necessary
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,316 @@
+            if in training mode
+        logger: a logger
+        modules_to_record (tuple): names of modules to be recorded.
+        ipu_model_cfg (dict): a dictionary contained train_split_edges,
```
```suggestion
        ipu_model_cfg (dict): a dictionary containes train_split_edges,
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,316 @@
+
+    Args:
+        model (nn.Module): The target model to be converted.
+        opts (dict): IPU options.
```
It would be helpful to describe the possible keys and values.

zhouzaida replied:
```suggestion
        opts (dict[str, poptorch.Options]): IPU options.
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,316 @@
+        mem_prop = {f'IPU{i}': availableMemoryProportion[i]
+                    for i in range(len(availableMemoryProportion))}
+        opts.setAvailableMemoryProportion(mem_prop)
+    if 'executionStrategy' in ipu_options:
```
```suggestion

    if 'executionStrategy' in ipu_options:
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,316 @@
+            opts.setExecutionStrategy(poptorch.ShardedExecution())
+        else:
+            raise NotImplementedError
+    if 'partialsType' in ipu_options:
```
```suggestion

    if 'partialsType' in ipu_options:
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+            key=None,
+            address='data'
+            ):
+        treeB = self._tree if treeB == 'ComplexDataManagerNone' else treeB
```
```suggestion
        if treeB == 'ComplexDataManagerNone:
            treeB = self._tree
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+    """Create a PopTorch inference model, from a PyTorch model, to run on IPU
+    hardware in inference mode.
+
+    .. note:: PopTorch makes a shallow copy of the model. Changes to the
+        parameters in the returned inference model affect the original model
+        and vice versa. However, primitive variable types are not synced: for
+        example calling ``model.eval()`` on the original model will not alter
+        the model returned by this function. You may need to call
+        ``model.eval()`` on your model before you call this function for
+        correct behavior.
```
```suggestion
    Note: 
        PopTorch makes a shallow copy of the model. Changes to the
        parameters in the returned inference model affect the original model
        and vice versa. However, primitive variable types are not synced: for
        example calling ``model.eval()`` on the original model will not alter
        the model returned by this function. You may need to call
        ``model.eval()`` on your model before you call this function for
        correct behavior.
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+        ``model.eval()`` on your model before you call this function for
+        correct behavior.
+
+    :param model: The PyTorch model to wrap.
+    :param options: The IPU specific options
+    :returns: The :py:class:`poptorch.PoplarExecutor` wrapper to use in place
+        of ``model``.
```
```suggestion
    Args:
        model (nn.Module or poptorch.PoplarExecutor): The PyTorch model to wrap.
        options (poptorch.Options, optional): The IPU specific options.
        logger (xxx): xxx.
   
   Returns:  
       The :class:`poptorch.PoplarExecutor` wrapper to use in place of ``model``.
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+def inferenceModel(model: Union['nn.Module', 'poptorch.PoplarExecutor'],
+                   options: Optional['poptorch.Options'] = None,
+                   logger=None
+                   ) -> 'poptorch.PoplarExecutor':
```
```suggestion
                   ) -> poptorch.PoplarExecutor:
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+                   options: Optional['poptorch.Options'] = None,
+                   logger=None
+                   ) -> 'poptorch.PoplarExecutor':
+    """Create a PopTorch inference model, from a PyTorch model, to run on IPU
```
```suggestion
    """Create a PopTorch inference model from a PyTorch model, to run on IPU
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+                  logger=None,
+                  modules_to_record=[]
+                  ) -> 'poptorch.PoplarExecutor':
+    """Create a PopTorch training model, from a PyTorch model, to run on IPU
```
```suggestion
    """Create a PopTorch training model from a PyTorch model, to run on IPU
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+        return getattr(self.executor, attr)
+
+
+def trainingModel(model: Union['nn.Module', 'poptorch.PoplarExecutor'],
+                  options: Optional['poptorch.Options'] = None,
+                  optimizer: Optional['torch.optim.Optimizer'] = None,
+                  logger=None,
+                  modules_to_record=[]
+                  ) -> 'poptorch.PoplarExecutor':
```
```suggestion
def get_training_model(model: Union[nn.Module, poptorch.PoplarExecutor],
                  options: Optional[poptorch.Options] = None,
                  optimizer: Optional[torch.optim.Optimizer] = None,
                  logger=None,
                  modules_to_record=[]
                  ) -> poptorch.PoplarExecutor:
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+        model returned by this function. You may need to call ``model.train()``
+        on your model before you call this function for correct behavior.
+
+    :param model: The PyTorch model to wrap.
+    :param options: The IPU specific options
+    :param optimizer: The optimizers to apply during \
+        training.
```
```suggestion
    Args:
        model (xxx): The PyTorch model to wrap.
        options (xxx): The IPU specific options
        optimizer (xxx): The optimizers to apply during training.
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+            :py:class:`poptorch.optim.RMSprop`. \
+            :py:class:`poptorch.optim.LAMB`.
+
+    :returns: The :py:class:`poptorch.PoplarExecutor` wrapper to use in place
+        of ``model``.
```
```suggestion
    returns:
        The :class:`poptorch.PoplarExecutor` wrapper to use in place
        of ``model``.
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+    :param optimizer: The optimizers to apply during \
+        training.
+
+        Supported PyTorch optimizers: ``optim.SGD``, ``optim.Adam``, \
+             ``optim.AdamW``, ``optim.RMSprop``.
+
+        Supported PopTorch optimizers: :py:class:`poptorch.optim.SGD`, \
+            :py:class:`poptorch.optim.Adam`, \
+            :py:class:`poptorch.optim.AdamW`, \
+            :py:class:`poptorch.optim.RMSprop`. \
+            :py:class:`poptorch.optim.LAMB`.
```
```suggestion
        - Supported PyTorch optimizers: ``optim.SGD``, ``optim.Adam``,
          ``optim.AdamW`` and ``optim.RMSprop``.

        - Supported PopTorch optimizers: ``poptorch.optim.SGD``,
           ``poptorch.optim.Adam``, ``poptorch.optim.AdamW``,
           ``poptorch.optim.RMSprop`` and ``poptorch.optim.LAMB``.
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+                f'not supported datatype:{str(_tree)}')
+
+
+class WrappedNet(nn.Module):
```
docstring is missing.

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+        return outputs
+
+
+class PoplarExecutorForMMCV(PoplarExecutor):
```
docstring is missing.

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+        else:
+            return a == b
+
+    def set_tree(self, _tree):
```
```suggestion
    def set_tree(self, tree):
```

zhouzaida replied:
docstring is missing.

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+        # tensors and basic python data types
+        if self._tree is not None:
+            if isinstance(_tree, torch.Tensor):
+                assert type(self._tree) == torch.Tensor, \
```
Can we use `isinstance(self._tree, torch.Tensor)`?

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+    def __init__(self, logger=None):
+        self.fixed_data_types = (int, str, float, np.ndarray, type(None))
+        self.warning = warnings.warn if logger is None else logger.warning
+        self.keys_of_changed_vals = []
```
It seems like the attribute is not used.

zhouzaida commented:
Some classes or methods miss docstring in `mmcv/runner/ipu/model_converter.py`

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+            strict=True,
+            key=None,
+            address='data'
+            ):
```
docstring is missing.

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+        else:
+            self._tree = _tree
+
+    def get_tree(self,):
+        return self._tree
```
Can we refactor the `get_tree` to the following format?
```python
@property
def tree(self):
    return self._tree

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,722 @@
+        # Update with a tree with the same structure
+        # but different values(tensors and basic python data types)
+        if isinstance(treeA, (tuple, list)):
+            for idx in range(len(treeA)):
```
```suggestion
            for idx, node in enumerate(treeA):
```

HuDi2018 committed:
Update mmcv/runner/ipu/model_converter.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/model_converter.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/model_converter.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/model_converter.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/model_converter.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/model_converter.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/model_converter.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/model_converter.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/runner/ipu/util.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
code review 3

HuDi2018 committed:
code review 3

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,324 @@
+    return model
+
+
+def recomputation_checkpoint(model: nn.Module, module_names)\
+     -> torch.utils.hooks.RemovableHandle:
```
```suggestion
def recomputation_checkpoint(model: nn.Module, module_names: list):
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,324 @@
+    for idx, (name, module) in enumerate(model.named_modules()):
+        if idx in spilt_edges_dict and name in spilt_edges_dict:
+            raise ValueError(
+                f'The same layer is referenced twice while doing model'
+                f' partition: idx is {idx} and name is {name}')
```
```suggestion
                'The same layer is referenced twice while doing model'
                f' partition: idx is {idx} and name is {name}')
```

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,144 @@
+       of IPUDataloader
+
+    Args:
+        model (pytorch.model)
```
```suggestion
        model (:obj:`torch.nn.Module`): The model to be run.
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,324 @@
+    Args:
+        model (nn.Module): The target model to be converted.
+        opts (dict[str, poptorch.Options]): IPU options, generated
+            by func: parse_ipu_options.
```
```suggestion
            by :func:`parse_ipu_options`.
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,324 @@
+        ipu_model_cfg=None,
+        fp16_cfg=None
+        ):
+    """Convert torch model to IPU model
```
```suggestion
    """Convert torch model to IPU model.
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,324 @@
+    some layers.
+
+    Args:
+        model (nn.Module): the target model to apply recomputation
+            checkpoint
+        module_names (list): model layer names
```
```suggestion
        model (nn.Module): The target model to apply recomputation
            checkpoint.
        module_names (list): Layer names of module.
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,324 @@
+
+def recomputation_checkpoint(model: nn.Module, module_names)\
+     -> torch.utils.hooks.RemovableHandle:
+    """Annotates the output of a module to be checkpointed instead of
+    recomputed
```
```suggestion
    """Annotates the output of a module to be checkpointed instead of
    recomputed.
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,324 @@
+            if in training mode
+        logger: a logger
+        modules_to_record (tuple): names of modules to be recorded.
+        ipu_model_cfg (dict): a dictionary contains train_split_edges,
+            train_ckpt_nodes, see details in funcations model_sharding and
+            recomputation_checkpoint
```
```suggestion
        ipu_model_cfg (dict): A dictionary contains train_split_edges and
            train_ckpt_nodes. See details in :func:`model_sharding` and
            :func:`recomputation_checkpoint` functions.
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,324 @@
+        ipu_model_cfg (dict): a dictionary contains train_split_edges,
+            train_ckpt_nodes, see details in funcations model_sharding and
+            recomputation_checkpoint
+        fp16_cfg (dict): config for IPU fp16 training, currently support
+            configs: `loss_scale`, `velocity_accum_type`, `accum_type`.
+            see details in
+            https://docs.graphcore.ai/projects/poptorch-user-guide/en/latest/index.html
```
```suggestion
        fp16_cfg (dict): Config for IPU fp16 training. Currently supports
            configs: `loss_scale`, `velocity_accum_type` and `accum_type`.
            See details in
            https://docs.graphcore.ai/projects/poptorch-user-guide/en/latest/index.html
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,324 @@
+            assert len(modules_to_record) == 0, \
+                'Feature alignment for multi-replica mode not implemented'
+
+    # split model into multi-IPUs if specified
+    assert len(ipu_model_cfg.get('eval_split_edges', [])) == 0,\
+        'Currently, BeginBlock can only be used once on the same model'
+    eval_model = model_sharding(copy.copy(model).eval(), ipu_model_cfg.get(
+        'eval_split_edges', []))
```
If `eval_split_edges` should not be set, is it necessary to call `model_shareding`?

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,324 @@
+        ipu_options (dict): A dictionary of ipu settings
+
+    Returns:
+        opts: training options and inference options of IPU in a
+            dictionary
```
```suggestion
        dict[str, poptorch.Options]: Training options and inference options of IPU.
        
```

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,144 @@
+
+
+class IPUBaseRunner(BaseRunner):
+    """An base runner for IPU.
```
```suggestion
    """A base runner for IPU.
```

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,144 @@
+        model (pytorch.model)
+        options (mmcv.Config, dict): Options that will be used to compile
+            and run the model.
+        modules_to_record (mmcv.Config, list): idx or name of modules which
```
```suggestion
        modules_to_record (mmcv.Config, list): Index or name of modules which
```

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,144 @@
+            static graph of model training or inference.
+        ipu_model_cfg (mmcv.Config, dict): config of model partition and
+            recomputing checkpoint
+        fp16_cfg (mmcv.Config): config for fp16 training
```
```suggestion
        fp16_cfg (mmcv.Config): Config for fp16 training.
```

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,144 @@
+        modules_to_record (mmcv.Config, list): idx or name of modules which
+            will be recorded for output. It is necessary to specify output for
+            static graph of model training or inference.
+        ipu_model_cfg (mmcv.Config, dict): config of model partition and
+            recomputing checkpoint
```
```suggestion
        ipu_model_cfg (mmcv.Config, dict): Config of model partition and
            recomputing checkpoint.
```

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,144 @@
+        fp16_cfg (mmcv.Config): config for fp16 training
+        batch_processor (callable): A callable method that process a data
+            batch. Should be None for IPU runner
+        kwargs: other kwargs please check Class base_runner.BaseRunner
```
```suggestion
        kwargs: Keyword arguments will be passed to ``base_runner.BaseRunner``.
```

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,144 @@
+        if isinstance(model, torch.nn.parallel.DataParallel):
+            raise TypeError(
+                'if you want to implement data parallelism '
+                'at the module level on IPU, '
```
```suggestion
                'at the module level on IPU, please'
```

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,144 @@
+                'at the module level on IPU, '
+                'use IPU option: replicationFactor')
+
+        ipu_options = ipu_options or {}
+        modules_to_record = modules_to_record or []
+        ipu_model_cfg = ipu_model_cfg or {}
+        super(IPUBaseRunner, self).__init__(model, **kwargs)
```
```suggestion
        # Need to add comments to explain how super() works here.
        super().__init__(model, **kwargs)
        
        ipu_options = ipu_options or {}
        modules_to_record = modules_to_record or []
        ipu_model_cfg = ipu_model_cfg or {}
```

zhouzaida commented on mmcv/runner/ipu_runner.py:
```diff
@@ -0,0 +1,144 @@
+
+        super().run(data_loaders, workflow, *args, **kwargs)
+
+    def get_ipu_opts(self, mode):
```
```suggestion
    def get_ipu_options(self, mode):
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,727 @@
+        # get a list of tensor from self._tree
+        target_tree = self._tree if target_tree is None else target_tree
+        tensors = []
+        if type(target_tree) == torch.Tensor:
```
Should we use `isinstance(target_tree, torch.Tensor)` ?

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,727 @@
+
+    def _get_tensors(self, tree, tensors):
+        if isinstance(tree, (tuple, list)):
+            for idx in range(len(tree)):
+                if isinstance(tree[idx], torch.Tensor):
+                    tensors.append(tree[idx])
+                else:
+                    self._get_tensors(tree[idx], tensors)
```
```suggestion
            for node in tree:
                if isinstance(node, torch.Tensor):
                    tensors.append(node)
                else:
                    self._get_tensors(node, tensors)
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,727 @@
+                else:
+                    self._get_tensors(tree[idx], tensors)
+        elif isinstance(tree, dict):
+            for k, v in tree.items():
+                if isinstance(v, torch.Tensor):
+                    tensors.append(tree[k])
+                else:
+                    self._get_tensors(tree[k], tensors)
```
```suggestion
            for k, v in tree.items():
                if isinstance(v, torch.Tensor):
                    tensors.append(v)
                else:
                    self._get_tensors(v, tensors)
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,727 @@
+                f'not supported datatype:{str(tree)}')
+
+    def set_tensors(self, tensors):
+        if type(self._tree) == torch.Tensor:
```
Same as above.

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,727 @@
+    def set_tensors(self, tensors):
+        if type(self._tree) == torch.Tensor:
+            assert len(tensors) == 1
+            assert type(tensors[0]) == torch.Tensor
```
Same as above.

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,727 @@
+                if isinstance(v, torch.Tensor):
+                    tree[k] = tensors.pop(0)
+                else:
+                    self._set_tensors(tree[k], tensors)
```
```suggestion
                    self._set_tensors(v, tensors)
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,727 @@
+                if isinstance(v, torch.Tensor):
+                    tree[k] = None
+                else:
+                    self._clean_tensors(tree[k])
```
```suggestion
                    self._clean_tensors(v)
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,727 @@
+            model,
+            inputs_tree_manager,
+            outputs_tree_manager,
+            modules_to_record=[],
```
It is not recommended to use mutable `[]` as a default value.

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,727 @@
+            inputs_tree_manager,
+            outputs_tree_manager,
+            modules_to_record=[],
+            hooked_features={}
```
Same as above.

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,727 @@
+
+    def forward(self, inputs_tuple):
+        # convert tuple back to kwargs
+        self.inputs_tree_manager.set_tensors(list(inputs_tuple))
```
Is it necessary to call `list(inputs_tuple)`?

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,727 @@
+        return input_output_hook
+
+    def forward(self, inputs_tuple):
+        # convert tuple back to kwargs
```
Briefly describe why inputs_tuple need to be converted to keyword arguments

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,727 @@
+    def update(
+            self,
+            treeA,
+            treeB='ComplexDataManagerNone',
```
What does the `ComplexDataManagerNone` mean?

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,727 @@
+                                key, address=new_address)
+        elif isinstance(treeA, self.fixed_data_types):
+            if not self.quick_mode:
+                is_equal = self.compare_fixed_type(treeA, treeB)
+                if strict:
+                    assert is_equal, 'all data except torch.Tensor '\
+                        'should be same, but data({}) is changed'.\
+                        format(address)
+                else:
+                    self.warning(f'find a non-torch.Tensor data({type(treeA)}) \
+                        changed, and the address is {address}')
```
```suggestion
                is_equal = 
                if strict:
                    if not self.compare_fixed_type(treeA, treeB):
                        raise ValueError(
                            'all data except torch.Tensor 'should be same, '
                            f'but data({address}) is changed'.)
                else:
                    self.warning(
                        f'find a non-torch.Tensor data({type(treeA)}) '
                        f'changed, and the address is {address}')
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,727 @@
+            eval_model, options=options['inference'], logger=logger)
+
+    @property
+    def executor(self,):
```
```suggestion
    def executor(self):
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,727 @@
+            assert not self.training
+
+    @property
+    def training(self,):
```
```suggestion
    def training(self):
```

HuDi2018 committed:
code review 4

HuDi2018 committed:
add some docstrings

HuDi2018 committed:
remove some whitespaces

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,324 @@
+        ipu_options (dict): A dictionary of ipu settings
+
+    Returns:
+        dict[str, poptorch.Options]: Training options and inference options
+            of IPU.
```
```suggestion
        dict[str, poptorch.Options]: Training options and inference options
        of IPU.
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,324 @@
+    if 'availableMemoryProportion' in ipu_options:
+        availableMemoryProportion = ipu_options.pop(
+            'availableMemoryProportion')
+        mem_prop = {f'IPU{i}': availableMemoryProportion[i]
+                    for i in range(len(availableMemoryProportion))}
```
```suggestion
        mem_props = {}
        for i, mem_prop in enumerate(availableMemoryProportion):
            mem_props[f'IPU{i}'] = mem_prop
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,324 @@
+        elif executionStrategy == 'ShardedExecution':
+            opts.setExecutionStrategy(poptorch.ShardedExecution())
+        else:
+            raise NotImplementedError
```
```suggestion
            raise NotImplementedError(
                f'executionStrategy should be "SameAsIpu" or "ShardedExecution", but got {execution_strategy}')
```

zhouzaida commented on mmcv/runner/ipu/util.py:
```diff
@@ -0,0 +1,324 @@
+        opts.setAvailableMemoryProportion(mem_prop)
+
+    if 'executionStrategy' in ipu_options:
+        executionStrategy = ipu_options.pop('executionStrategy')
```
```suggestion
        execution_strategy = ipu_options.pop('executionStrategy')
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,809 @@
+
+
+class DictArgsParser(ArgsParser):
+    """A helper class for handling model input.
+    Args:
+        inputs (list): Inputs of model.
+    """
```
```suggestion
    """A helper class for handling model input.
    
    Args:
        inputs (list): Inputs of model.
    """
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,809 @@
+            self.outputs_tree_manager.quick()
+
+        # parser args for first iter
+        self._args_parser = DictArgsParser(
+            {'args': inputs_tuple}) if self._args_parser is None\
+            else self._args_parser
```
```suggestion
        if self._args_parser is None:
            self._args_parser = DictArgsParser({'args': inputs_tuple})
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,809 @@
+        # according to the same order
+        self.outputs_tree_manager.set_tensors(plain_outputs)
+        # get the real output dictionary from self.outputs_tree_manager
+        output_dic = self.outputs_tree_manager.tree
```
```suggestion
        output_dict = self.outputs_tree_manager.tree
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,809 @@
+
+
+class TrainEvalModel:
+    """A class maintaining training PoplarExecutorForMMCV and inference
+        PoplarExecutorForMMCV.
```
```suggestion
    """A class maintaining training PoplarExecutorForMMCV and inference
    PoplarExecutorForMMCV.
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,809 @@
+        PoplarExecutorForMMCV.
+
+    Args:
+        train_model (:obj:`nn.Module`): The training model to be compiled.
```
```suggestion
        train_model (:obj:`nn.Module`): The training model to be compiled.
            ``train_model`` can be None if only executing validation.
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,809 @@
+        e.g. :class:`Dropout`, :class:`BatchNorm`, etc.
+
+        Args:
+            mode (bool): whether to set training mode (``True``) or evaluation
+                         mode (``False``). Default: ``True``.
```
```suggestion
            mode (bool): Whether to set training mode (``True``) or evaluation
                mode (``False``). Default: ``True``.
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,809 @@
+        """
+        if not isinstance(mode, bool):
+            raise ValueError('training mode is expected to be boolean')
+        if self._train_executor is None and mode is True:
```
```suggestion
        if self._train_executor is None and mode:
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,809 @@
+    def train_step(self, data, optimizer=None, **kwargs):
+        assert self.training, 'not supported train_step on eval mode'
+        hooked_features_cpu = {}
+        if self._train_executor.isCompiled() \
+                and self._train_executor.compare_with_cpu:
```
```suggestion
        if (self._train_executor.isCompiled() and
            self._train_executor.compare_with_cpu):
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,809 @@
+            self._train_executor.model.train_step(data, optimizer, **kwargs)
+            hooked_features_cpu = {**(self._train_executor.hooked_features)}
+        result = self._train_executor.train_step(data, optimizer, **kwargs)
+        if self._train_executor.isCompiled() \
+            and self._train_executor.compare_with_cpu \
+                and len(hooked_features_cpu) > 0:
```
```suggestion
        if (self._train_executor.isCompiled() and
            self._train_executor.compare_with_cpu and
            len(hooked_features_cpu) > 0):
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,809 @@
+        """
+        return self.train(False)
+
+    def compare_data_between_ipu_and_cpu(
```
This method just makes comparisons. Should it assert an error?

HumberMe replied:
Occasionally, there is inconsistency between the calculation of the IPU and the calculation of the CPU, but whether it is reasonable or not needs to be judged by engineers.

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,809 @@
+        self.logger = logger
+        self.hooked_features = {}
+        self.hooked_features_ipu = {}
+        self.compare_with_cpu = False if modules_to_record is None else True
```
Need to describe why we need `self.compare_with_cpu`.

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,809 @@
+                2. Torch.Tensor data shape changed.
+            address: Record the address of current data to be updated.
+        """
+        if treeB == ComplexDataManagerNone:
```
```suggestion
        if treeB is ComplexDataManagerNone:
```

HuDi2018 committed:
test coverage rate reaches 100%

HuDi2018 committed:
little change to mmcv/runner/ipu/model_converter.py:compare_data_between_ipu_and_cpu

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,814 @@
+    def __init__(self, logger=None):
+        self.fixed_data_types = (int, str, float, np.ndarray, type(None))
+        self.warning = warnings.warn if logger is None else logger.warning
+        self.quick_mode = False
```
Please add comments to describe the attribute.

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,814 @@
+                    assert isinstance(treeA[k], type(treeB[k])),\
+                        f'data structure changed: {new_address}'
+                if isinstance(v, torch.Tensor):
+                    treeB[k] = treeA[k]
```
```suggestion
                    treeB[k] = v
```

zhouzaida commented on mmcv/runner/ipu/model_converter.py:
```diff
@@ -0,0 +1,814 @@
+ComplexDataManagerNone = object()
+
+
+class ComplexDataManager:
```
It would be helpful if an example is provided to show how input is converted to the target data structure.

HuDi2018 committed:
code review 5

HuDi2018 committed:
code review 5

HuDi2018 committed:
adjust code structure

zhouzaida commented:
related pr: https://github.com/open-mmlab/mmclassification/pull/723

ZwwWayne commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,829 @@
+    training/inference model.
+
+    Args:
+        model (:obj:`nn.Module`): The model to be run.
```
to be run. -> to run.

ZwwWayne commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,829 @@
+        self.inputs_tree_manager = ComplexDataManager(logger=logger)
+        self.outputs_tree_manager = ComplexDataManager(logger=logger)
+        self.logger = logger
+        self.hooked_features = {}
```
might some explanation of difference between hooked_features_ipu and hooked_features and what they are used for.

ZwwWayne commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,829 @@
+
+        # split output_dict into hooked_features_ipu
+        # and output of the torch model
+        mmcv_model_output = {}
```
according to the comments in lines 510&511, it might be better to rename it to torch_model_output

ZwwWayne commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,829 @@
+            Module: self
+        """
+        if not isinstance(mode, bool):
+            raise ValueError('training mode is expected to be boolean')
```
may tell what type is got to help debug

ZwwWayne commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,328 @@
+        if callable(options_node):
+            options_node(cfg)
+        else:
+            error_msg = f'opts_node type {type(options_node)} not supported'
```
opts_node or options_node?

ZwwWayne commented on mmcv/device/ipu/runner.py:
```diff
@@ -0,0 +1,133 @@
+                modules_to_record=modules_to_record,
+                ipu_model_cfg=ipu_model_cfg, fp16_cfg=fp16_cfg)
+        else:
+            raise NotImplementedError('cpu mode on IPURunner not supported')
```
IPURunner not supported -> IPURunner is not supported

zhouzaida commented on mmcv/device/ipu/dataloader.py:
```diff
@@ -0,0 +1,152 @@
+        return default_collate(batch)
+
+
+class IPUDataloader(poptorch.DataLoader):
```
```suggestion
class IPUDataLoader(poptorch.DataLoader):
```

HuDi2018 committed:
Update docs/en/get_started/build.md

Co-authored-by: Wenwei Zhang <40779233+ZwwWayne@users.noreply.github.com>

HuDi2018 committed:
Update docs/en/get_started/build.md

Co-authored-by: Wenwei Zhang <40779233+ZwwWayne@users.noreply.github.com>

HuDi2018 committed:
Update docs/en/get_started/build.md

Co-authored-by: Wenwei Zhang <40779233+ZwwWayne@users.noreply.github.com>

HuDi2018 committed:
Update mmcv/device/ipu/dataloader.py

Co-authored-by: Zaida Zhou <58739961+zhouzaida@users.noreply.github.com>

HuDi2018 committed:
little change to tests

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,328 @@
+                f', but got {execution_strategy}')
+
+    if 'partialsType' in ipu_options:
+        partialsType = ipu_options.pop('partialsType')
+        opts.Precision.setPartialsType(
+            getattr(torch, partialsType))  # half or float
```
```suggestion
        partials_type = ipu_options.pop('partialsType')
        opts.Precision.setPartialsType(
            getattr(torch, partials_type))  # half or float
```

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,328 @@
+    opts = poptorch.Options()
+
+    if 'availableMemoryProportion' in ipu_options:
+        availableMemoryProportion = ipu_options.pop(
+            'availableMemoryProportion')
+        mem_props = {}
+        for i, mem_prop in enumerate(availableMemoryProportion):
```
```suggestion
        available_memory_proportion = ipu_options.pop(
            'availableMemoryProportion')
        mem_props = {}
        for i, mem_prop in enumerate(available_memory_proportion):
```

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,328 @@
+def _parse_ipu_options(ipu_options):
+    # If it cannot be directly assigned, use if statement to parse it,
+    # and if it can be directly assigned, use _options_assigner to assign
+    opts = poptorch.Options()
```
```suggestion
    options = poptorch.Options()
```

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,328 @@
+    # overwrite default ipu options with specified eval cfgs
+    inference_ipu_options = {**ipu_options, **eval_cfgs}
+
+    opts = {'training': _parse_ipu_options(training_ipu_options),
```
```suggestion
    options = {'training': _parse_ipu_options(training_ipu_options),
```

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,328 @@
+
+def ipu_model_wrapper(
+        model,
+        opts,
```
```suggestion
        options,
```

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,328 @@
+
+
+def parse_ipu_options(ipu_options):
+    """parse dictionary to ipu options.
```
```suggestion
    """Parse dictionary to ipu options.
```

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,328 @@
+        raise NotImplementedError(error_msg)
+
+
+def parse_ipu_options(ipu_options):
```
```suggestion
def parse_ipu_options(options):
```

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,328 @@
+
+    Args:
+        model (nn.Module): The target model to be converted.
+        opts (dict[str, poptorch.Options]): IPU options, generated
```
```suggestion
        ipu_options (dict[str, poptorch.Options]): IPU options, generated
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,832 @@
+            super().attachToDevice()
+
+
+def compare_feat(featA, featB, rtol=1e-3, atol=1e-5):
+    """Align data between two activations or weights."""
+    try:
+        np.testing.assert_allclose(featA, featB, rtol=rtol, atol=atol)
+    except AssertionError as e:
+        print(e)
```
```suggestion
def compare_tensor(input, other, rtol=1e-3, atol=1e-5):
    """Align data between two activations or weights."""
    try:
        torch.allclose(input, other, rtol=rtol, atol=atol)
    except AssertionError as e:
        print(e)
```

HumberMe replied:
torch.allclose will not raise exception if two tensors are not close

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,832 @@
+        if (self._train_executor.isCompiled() and
+            self._train_executor.compare_with_cpu):
+            self.copyWeightsToHost()
+            self._train_executor.model.train_step(data, optimizer, **kwargs)
+            hooked_features_cpu = {**(self._train_executor.hooked_features)}
+        result = self._train_executor.train_step(data, optimizer, **kwargs)
```
briefly describe the difference between `self._train_executor.model.train_step(data, optimizer, **kwargs)` and `self._train_executor.train_step(data, optimizer, **kwargs)`

HuDi2018 committed:
code review 6

HuDi2018 committed:
refine some errors

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,333 @@
+        of IPU.
+    """
+    # set ipu options for inference and training by config
+    train_cfgs = options.pop('train_cfgs', {})
+    eval_cfgs = options.pop('eval_cfgs', {})
```
```suggestion
    train_cfg = options.pop('train_cfg', {})
    eval_cfg = options.pop('eval_cfg', {})
```

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,333 @@
+        raise NotImplementedError(error_msg)
+
+
+def parse_ipu_options(options):
```
```suggestion
def cast_to_options(cfg):
```

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,333 @@
+    return ipu_options
+
+
+def _parse_ipu_options(options_dict):
```
```suggestion
def _cast_to_options(cfg):
```

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,333 @@
+    Args:
+        model (nn.Module): The target model to be converted.
+        options (dict[str, poptorch.Options]): IPU options, generated
+            by :func:`parse_ipu_options`.
```
```suggestion
            by :func:`vast_to_options`.
```

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,333 @@
+            by :func:`parse_ipu_options`.
+        optimizer (:obj:`torch.optim.Optimizer`, optional): torch
+            optimizer, necessary if in training mode
+        logger: a logger
```
```suggestion
        logger (:obj:`logging.Logger`): Logger used during training.
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,835 @@
+        """Update recorded complex data in-place.
+
+        Args:
+            treeA (list or dictionary or tuple): New complex data.
```
```suggestion
            treeA (list or dict or tuple): New complex data.
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,835 @@
+
+        Args:
+            treeA (list or dictionary or tuple): New complex data.
+            treeB (list or dictionary or tuple): Complex data to update,
```
```suggestion
            treeB (list or dict or tuple): Complex data to update,
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,835 @@
+                when the following conditions occur:
+                1. Non-torch.Tensor data changed.
+                2. Torch.Tensor data shape changed.
+            address: Record the address of current data to be updated.
```
```suggestion
            address (str): Record the address of current data to be updated.
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,835 @@
+                f'not supported datatype:{type(treeA)}, address is {address}')
+
+    def get_tensors(self, target_tree=None):
+        """Collect torch.Tensor data from self.tree to a tuple and return."""
```
```suggestion
        """Collect torch.Tensor data from self.tree to a list and return."""
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,835 @@
+
+    def forward_train(self, kwargs):
+        optimizer = kwargs.pop('optimizer')
+        data = kwargs
+        outputs = self.train_step(data, optimizer)
```
```suggestion
        outputs = self.train_step(kwargs, optimizer)
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,835 @@
+        self.outputs_tree_manager = HierarchicalData(logger=logger)
+        self.logger = logger
+        # the features calculated by CPU
+        self.hooked_features = {}
```
```suggestion
        self.hooked_features_cpu = {}
```

HuDi2018 committed:
cfgs to cfg

HuDi2018 committed:
parse_ipu_options to cast_to_options

HuDi2018 committed:
hooked_features to hooked_features_cpu

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,834 @@
+
+
+class WrappedNet(nn.Module):
+    """A net wrapper for model convertion.
```
```suggestion
    """A net wrapper for model conversion.
```

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,333 @@
+        raise NotImplementedError(error_msg)
+
+
+def cast_to_options(options):
```
Suggest rename `options` to `cfg`. The `options` is an instance of `poptorch.Options` and `cfg` is a dict to construct `options`.

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,333 @@
+    return ipu_options
+
+
+def _cast_to_options(options_dict):
```
same as above

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,333 @@
+
+    # set model partition
+    if optimizer is None:
+        train_model = None
```
```suggestion
        training_model = None
```

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,333 @@
+        module_names (list): Layer names of module.
+    """
+    def recompute_outputs(module, inputs, outputs):
+        if type(outputs) is tuple:
```
Should we use `isinstance(outputs, tuple)`?

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,834 @@
+
+        Args:
+            treeA (list or dict or tuple): New complex data.
+            treeB (list or dict or tuple): Complex data to update,
+                if not entered here, self.tree will be updated then.
```
```suggestion
            treeB (list or dict or tuple): Complex data to update.
                if not specified, self.tree will be updated then.
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,834 @@
+                when the following conditions occur:
+                1. Non-torch.Tensor data changed.
+                2. Torch.Tensor data shape changed.
+            address (str): Record the address of current data to be updated.
```
```suggestion
            address (str): Record the address of current data to be updated.
                Default: 'data'.
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,834 @@
+        if isinstance(tree, (tuple, list)):
+            for idx in range(len(tree)):
+                if isinstance(tree[idx], torch.Tensor):
+                    tree[idx] = tensors.pop(0)
```
If tree is tuple, tree[idx] will throw an exception.

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,834 @@
+        if isinstance(tree, (tuple, list)):
+            for idx in range(len(tree)):
+                if isinstance(tree[idx], torch.Tensor):
+                    tree[idx] = None
```
same as above.

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,834 @@
+        self.outputs_tree_manager = outputs_tree_manager
+        self.training = model.training
+        # Register a hook function to capture the intermediate features
+        # generated by the network to align the outputs between ipu and cpu
```
Need to add extra comments to explain why we need to align them.

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,834 @@
+        self.training = model.training
+        # Register a hook function to capture the intermediate features
+        # generated by the network to align the outputs between ipu and cpu
+        self.hooked_features_cpu = hooked_features_cpu
```
suggest renaming `hooked_features_cpu` to `inter_ouputs_in_cpu`.

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,834 @@
+        return outputs
+
+
+class PoplarExecutorForMMCV(PoplarExecutor):
```
```suggestion
class MMPoplarExecutor(PoplarExecutor):
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,834 @@
+        self.outputs_tree_manager = HierarchicalData(logger=logger)
+        self.logger = logger
+        # the features calculated by CPU
+        self.hooked_features_cpu = {}
```
```suggestion
        self.inter_outputs_in_cpu = {}
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,834 @@
+        # the features calculated by CPU
+        self.hooked_features_cpu = {}
+        # the features calculated by IPU
+        self.hooked_features_ipu = {}
```
```suggestion
        self.inter_outputs_in_ipu = {}
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,834 @@
+
+    def compare_data_between_ipu_and_cpu(
+            self,
+            hooked_features_cpu,
+            hooked_features_ipu):
```
```suggestion
            inter_outputs_from_cpu,
            inter_outputs_from_ipu):
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,834 @@
+    # merge train_step(train) and __call__(eval) together
+    def train_step(self, data, optimizer=None, **kwargs):
+        assert self.training, 'not supported train_step on eval mode'
+        hooked_features_cpu = {}
```
same as above

zhouzaida commented on mmcv/device/ipu/utils.py:
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+
+
+def get_ipu_mode():
```
Consider renaming `get_ipu_mode` to `is_ipu` and `IPU_MODE` to `IS_IPU`.

zhouzaida commented on tests/test_ipu/test_complexdatamanager.py:
```diff
@@ -0,0 +1,78 @@
+            'b': np.random.rand(3, 4),
+            'c': [1, 'asd']}}
+
+    cdm = HierarchicalData(logging.getLogger())
```
```suggestion
    hd = HierarchicalData(logging.getLogger())
```

zhouzaida commented on mmcv/device/ipu/dataloader.py:
```diff
@@ -0,0 +1,152 @@
+        """
+        # lazy init: sometimes, we cannot get IPU options when build data
+        #            loader
+        self.kwargs = {'options': options,
```
Is it necessary to contain 'options' here because 'options' will be overwritten in `init`?

HuDi2018 committed:
options to cfg

HuDi2018 committed:
commit before hooked_features_ipu to inter_ouputs_in_ipu

HuDi2018 committed:
hooked_features_ipu to inter_ouputs_in_ipu

HuDi2018 committed:
code review 7

HuDi2018 committed:
init ipu in mmcv/__init__.py

HuDi2018 committed:
move test_complexdatamanager.py to test_hierarchicaldata.py

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,853 @@
+        # generated by the network to align the outputs between ipu and cpu
+        # Used to confirm whether the implementation of CPU is consistent
+        # with the implementation of IPU
+        self.inter_ouputs_in_cpu = inter_ouputs_in_cpu
```
```suggestion
        self.inter_outputs_in_cpu = inter_outputs_in_cpu
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,853 @@
+        # temporarily use this function to fix the problem
+        return self._training  # comes from self.model._training
+
+    @auto_fp16(supported_types=(PoplarExecutor,))
```
Should it be `nn.Module, PoplarExecutor`?

HumberMe replied:
it decorate PoplarExecutor's function, so just PoplarExecutor is okay

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,853 @@
+            raise NotImplementedError(
+                f'not supported datatype:{type(tree)}')
+
+    def set_tensors(self, tensors):
```
```suggestion
    def to_tree(self, tensors):
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,853 @@
+            raise NotImplementedError(
+                f'not supported datatype:{type(treeA)}, address is {address}')
+
+    def get_tensors(self, target_tree=None):
```
```suggestion
    def tree_to_tuple(self, target_tree=None):
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,853 @@
+            raise NotImplementedError(
+                f'not supported datatype:{type(tree)}')
+
+    def clean_tensors(self):
```
```suggestion
    def clean_tree(self):
```

zhouzaida commented on mmcv/device/ipu/model_converter.py:
```diff
@@ -0,0 +1,853 @@
+
+        return output_dict
+
+    def detachFromDevice(self):
+        if self.isCompiled() and self._is_attached:
+            super().detachFromDevice()
+
+    def attachToDevice(self):
+        if self.isCompiled() and not self._is_attached:
+            super().attachToDevice()
```
Is it necessary to overwrite the two methods?

HumberMe replied:
it is necessary

HuDi2018 committed:
change names of hierarchicaldata

HuDi2018 committed:
move ipu intialization from mmcv/__init__.py to mmcv/runner/__init__.py

HuDi2018 committed:
remove redundancy codes

zhouzaida commented:
move tests/test_ipu to tests/test_device/test_ipu

zhouzaida commented on mmcv/device/ipu/runner.py:
```diff
@@ -0,0 +1,138 @@
+from mmcv.runner import (BaseRunner, IterBasedRunner,
+                         EpochBasedRunner, RUNNERS, HOOKS)
+
+from . import IS_IPU
```
```suggestion
from .utils import IS_IPU
```

HuDi2018 committed:
fix import problem

HuDi2018 committed:
add docs, adjust import things

zhouzaida commented on mmcv/device/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from . import ipu
```
```suggestion
from .utils import ipu
```

zhouzaida commented on mmcv/device/ipu/runner.py:
```diff
@@ -0,0 +1,138 @@
+    def __init__(
+            self,
+            model,
+            ipu_options=None,
```
```suggestion
            options_cfg=None,
```

zhouzaida commented on mmcv/device/ipu/runner.py:
```diff
@@ -0,0 +1,138 @@
+        assert hasattr(model, 'train_step') and batch_processor is None,\
+            'only support model with train_step'
+
+        if ipu_options is None:
+            ipu_options = {}
```
```suggestion
        if options_cfg is None:
            options_cfg = {}
```

zhouzaida commented on tests/test_device/test_ipu/test_ipu_dataloder.py:
```diff
@@ -0,0 +1,45 @@
+
+
+@skip_no_ipu
+def test_collate_ipu():
```
```suggestion
def test_ipu_collate():
```

Commit:
remove engine_cache in .gitignore

Commit:
refine dataloder test

Commit:
changed by pre-commit

Commit:
solve some bugs

zhouzaida commented on mmcv/device/ipu/runner.py:
```diff
@@ -0,0 +1,140 @@
+
+        super().run(data_loaders, workflow, *args, **kwargs)
+
+    def get_ipu_options(self, mode):
```
```suggestion
    def get_options(self, mode):
```

zhouzaida commented on mmcv/device/ipu/runner.py:
```diff
@@ -0,0 +1,140 @@
+
+        # process options of ipu
+        if IS_IPU:
+            self.ipu_options = cast_to_options(options_cfg)
```
```suggestion
            self.options = cast_to_options(options_cfg)
```

zhouzaida commented:
`tests/test_device/test_ipu/test_hierarchicaldatamanager.py` is repeated with `tests/test_device/test_ipu/test_hierarchicaldata.py`.

Commit:
123

Commit:
merge with master

zhouzaida commented on mmcv/device/ipu/hook_wrapper.py:
```diff
@@ -0,0 +1,105 @@
+def wrap_lr_updater_hook(lr_hook_class):
+    """A wrapper function to wrap any subclass of LrUpdaterHook.
+
+    IPU needs extra operations to upload optimizer settings. this wrapper will
```
```suggestion
    IPU needs extra operations to upload optimizer settings. This wrapper will
```

zhouzaida commented on mmcv/runner/__init__.py:
```diff
@@ -1,4 +1,5 @@
 # Copyright (c) OpenMMLab. All rights reserved.
+# initialize ipu to registor ipu runner to RUNNERS
```
This line should be moved to the front of line 45.

zhouzaida commented on tests/test_device/test_ipu/test_hierarchicaldata.py:
```diff
@@ -0,0 +1,91 @@
+    hd = HierarchicalDataManager(logging.getLogger())
+    hd.record_hierarchical_data(hierarchical_data_sample)
+    tensors = hd.get_all_tensors()
+    tensors[0].add_(1)
```
same as above

zhouzaida commented on tests/test_device/test_ipu/test_hierarchicaldata.py:
```diff
@@ -0,0 +1,91 @@
+    data = hd.data
+    data['c'].data['a'].sub_(1)
+    hd.record_hierarchical_data(data)
+    tensors = hd.get_all_tensors()
```
same as above

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,337 @@
+from .model_converter import TrainEvalModel
+
+
+def build_from_cfg_with_wrapper(cfg,
```
move this function to utis.py

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,337 @@
+        raise type(e)(f'{wrapped_obj_cls.__name__}: {e}')
+
+
+def _options_assigner(cfg, options_node):
```
move this function to utis.py

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,337 @@
+        raise NotImplementedError(error_msg)
+
+
+def cast_to_options(cfg):
```
move this function to utis.py

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,337 @@
+    return ipu_options
+
+
+def _cast_to_options(cfg):
```
move this function to utis.py

zhouzaida commented on tests/test_device/test_ipu/test_hierarchicaldata.py:
```diff
@@ -0,0 +1,91 @@
+    hd = HierarchicalDataManager(logging.getLogger())
+    hd.record_hierarchical_data(single_tensor)
+    hd.record_hierarchical_data(torch.rand(3, 4))
+    tensors = hd.get_all_tensors()
```
same as above

zhouzaida commented on .github/workflows/build.yml:
```diff
@@ -45,7 +45,7 @@ jobs:
         run: |
           pip install -r requirements/test.txt
-          pytest tests/ --ignore=tests/test_runner --ignore=tests/test_optimizer.py --ignore=tests/test_cnn --ignore=tests/test_parallel.py --ignore=tests/test_ops --ignore=tests/test_load_model_zoo.py --ignore=tests/test_utils/test_logging.py --ignore=tests/test_image/test_io.py --ignore=tests/test_utils/test_registry.py --ignore=tests/test_utils/test_parrots_jit.py --ignore=tests/test_utils/tes ...
+          pytest tests/ --ignore=tests/test_runner --ignore=tests/test_ipu --ignore=tests/test_optimizer.py --ignore=tests/test_cnn --ignore=tests/test_parallel.py --ignore=tests/test_ops --ignore=tests/test_load_model_zoo.py --ignore=tests/test_utils/test_logging.py --ignore=tests/test_image/test_io.py --ignore=tests/test_utils/test_registry.py --ignore=tests/test_utils/test_parrots_jit.py --ign ...
```
```suggestion
          pytest tests/ --ignore=tests/test_runner --ignore=tests/test_device/test_ipu --ignore=tests/test_optimizer.py --ignore=tests/test_cnn --ignore=tests/test_parallel.py --ignore=tests/test_ops --ignore=tests/test_load_model_zoo.py --ignore=tests/test_utils/test_logging.py --ignore=tests/test_image/test_io.py --ignore=tests/test_utils/test_registry.py --ignore=tests/test_utils/test_parrots_jit.py --ignore=tests/test_utils/test_trace.py --ignore=tests/test_utils/test_hub.py
```

Commit:
adjust code structure

zhouzaida commented:
`docs/en/api.rst` and `docs/zh_cn/api.rst` can be temporarily removed becourse readthedocs can not generate the API for mmcv/device without IPU environment.

Commit:
code review 8

zhouzaida commented on mmcv/device/ipu/hierarchical_data_manager.py:
```diff
@@ -0,0 +1,239 @@
+    def data(self):
+        return self._data
+
+    def update(self,
```
```suggestion
    def update_hierarchical_data(self,
```

zhouzaida commented on mmcv/device/ipu/hierarchical_data_manager.py:
```diff
@@ -0,0 +1,239 @@
+        self.warning = warnings.warn if logger is None else logger.warning
+        # enable or disable input data's shape and value check
+        self.quick_mode = False
+        self._data = None
```
```suggestion
        self._hierarchical_data = None
```

zhouzaida commented on mmcv/device/ipu/hierarchical_data_manager.py:
```diff
@@ -0,0 +1,239 @@
+            self._data = data
+
+    @property
+    def data(self):
+        return self._data
```
```suggestion
    def hierarchical_data(self):
        return self._hierarchical_data
```

zhouzaida commented on mmcv/device/ipu/hierarchical_data_manager.py:
```diff
@@ -0,0 +1,239 @@
+               dataB=HierarchicalDataNone,
+               strict=True,
+               address='data'):
+        """Update recorded complex data in-place.
```
```suggestion
        """Update dataB with dataA in-place.
```

zhouzaida commented on mmcv/device/ipu/hierarchical_data_manager.py:
```diff
@@ -0,0 +1,239 @@
+            raise NotImplementedError(
+                f'not supported datatype:{type(dataA)}, address is {address}')
+
+    def get_all_tensors(self, hierarchical_data=None):
```
```suggestion
    def collect_all_tensors(self, hierarchical_data=None):
```

zhouzaida commented on mmcv/device/ipu/hierarchical_data_manager.py:
```diff
@@ -0,0 +1,239 @@
+            self._get_tensors(hierarchical_data, tensors)
+        return tensors
+
+    def _get_tensors(self, data, tensors):
```
```suggestion
    def _collect_tensors(self, data, tensors):
```

zhouzaida replied:
```suggestion
    def _collect_tensors(self, data, tensors):
        """Recursively collect tensors from data."""
```

zhouzaida commented on tests/test_device/test_ipu/test_ipu_dataloder.py:
```diff
@@ -0,0 +1,72 @@
+    }
+    batch = [batch1, batch0]
+    results = collate(batch)
+    tensor_shape = torch.Tensor(list(results['tensor'].shape))
+    arr_shape = torch.Tensor(list(results['arr'].shape))
+    assert torch.equal(torch.Tensor([2, 3, 4, 5]), tensor_shape)
+    assert torch.equal(torch.Tensor([2, 3, 4, 5, 6]), arr_shape)
```
```suggestion
    assert results['tensor'].shape == (2, 3, 4, 5)
    assert results['arr'].shape == (2, 3, 4, 5, 6)
```

HumberMe replied:
seems like it will return false

zhouzaida commented on tests/test_device/test_ipu/test_ipu_dataloder.py:
```diff
@@ -0,0 +1,72 @@
+    for data in results['data_list']:
+        for tensor in data:
+            assert not isinstance(tensor, DataContainer)
+            data_in_list_shape = torch.Tensor(list(tensor.shape))
+            assert torch.equal(torch.Tensor([2]), data_in_list_shape)
```
```suggestion
            assert tensor.shape == (2,)
```

zhouzaida commented on tests/test_device/test_ipu/test_ipu_hooks.py:
```diff
@@ -0,0 +1,134 @@
+        }
+    }
+
+    temp_root = tempfile.gettempdir()
```
This temporary directory should be removed after testing or use the tmp_path fixture from pytest (https://docs.pytest.org/en/6.2.x/tmpdir.html#the-tmp-path-fixture)

zhouzaida commented on tests/test_device/test_ipu/test_ipu_hooks.py:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    temp_root = tempfile.gettempdir()
+    dir_name = ''.join(
+        [random.choice(string.ascii_letters) for _ in range(10)])
```
Why do we need to generate a random directory name?

HumberMe replied:
no idea, just copy from tests/test_runner/test_runner.py p49

zhouzaida commented on tests/test_device/test_ipu/test_ipu_hooks.py:
```diff
@@ -0,0 +1,134 @@
+
+    # test fp16 optimizer hook
+    lr_config = dict(policy='step', step=[100, 150])
+    optimizer_config = dict(grad_clip=dict(max_norm=2), )
```
```suggestion
    optimizer_config = dict(grad_clip=dict(max_norm=2))
```

zhouzaida commented on mmcv/device/ipu/hook_wrapper.py:
```diff
@@ -0,0 +1,105 @@
+                raise NotImplementedError(
+                    'IPU mode supports single scale currently')
+            else:
+                raise ValueError('loss_scale must be of type float, dict, or '
+                                 f'"dynamic", but got {loss_scale}')
```
```suggestion
                raise ValueError(f'loss_scale should be float, but got {loss_scale} ')
```

zhouzaida commented on tests/test_device/test_ipu/test_ipu_hooks.py:
```diff
@@ -0,0 +1,134 @@
+
+
+@skip_no_ipu
+def test_optimizerhook():
```
```suggestion
def test_ipu_hook_wrapper():
```

zhouzaida commented on mmcv/device/ipu/ipu_wrapper.py:
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+
+
+def is_ipu():
+    try:
+        import poptorch  # noqa: E261, F401
+        IS_IPU = True
+    except ImportError:
+        IS_IPU = False
+    return IS_IPU
+
+
+IS_IPU = is_ipu()
```
Move this function and `IS_IPU` to utils.py

HumberMe replied:
if put IS_IPU to utils.py, all the codes of utils.py will be placed in a if block(if IS_IPU)

zhouzaida commented on tests/test_device/test_ipu/test_ipu_hooks.py:
```diff
@@ -0,0 +1,134 @@
+    dummy_runner.register_training_hooks(
+        lr_config=lr_config, optimizer_config=None, timer_config=None)
+
+    # test _set_lr()
```
It is better to test whether the learning rate is set as expected when calling `dummy_runner.run()`.

HumberMe replied:
can only get learning rate from pytorch optimizer(cpu), the data on IPU cannot get.
our sdk will have additional tests to check if the optimizer update is working normal

zhouzaida replied:
got it

zhouzaida commented on mmcv/device/ipu/model_wrapper.py:
```diff
@@ -0,0 +1,721 @@
+from .utils import compare_tensor, model_sharding, recomputation_checkpoint
+
+
+class DictArgsParser(ArgsParser):
```
Also, move this class to utils.py

zhouzaida commented on tests/test_device/test_ipu/test_ipu_runner.py:
```diff
@@ -0,0 +1,131 @@
+@skip_no_ipu
+def test_build_runner():
+    # __init__
+    temp_root = tempfile.gettempdir()
```
This temporary directory should be removed after testing or use the tmp_path fixture from pytest (https://docs.pytest.org/en/6.2.x/tmpdir.html#the-tmp-path-fixture)

zhouzaida commented on mmcv/device/ipu/utils.py:
```diff
@@ -0,0 +1,244 @@
+    return options
+
+
+def model_sharding(model, split_edges):
```
`model_sharding` and `recomputation_checkpoint` should be moved to `model_wrapper.py`.

zhouzaida commented on tests/test_device/test_ipu/test_ipu_utils.py:
```diff
@@ -0,0 +1,181 @@
+    ]
+
+    with pytest.raises(ValueError, match='The same layer is referenced'):
+        model_sharding(model, split_edges)
```
A successful unit test should be added.

zhouzaida commented on tests/test_device/test_ipu/test_ipu_utils.py:
```diff
@@ -0,0 +1,181 @@
+        _options_cfg = copy.deepcopy(options_cfg)
+        _options_cfg['train_cfg']['Precision'] = {'autocast_policy': 123}
+        cast_to_options(_options_cfg)
+
```
Is it possible to test `recomputation_checkpoint`?

HumberMe replied:
can only check by profile manully

zhouzaida replied:
ok

zhouzaida commented on tests/test_device/test_ipu/test_ipu_utils.py:
```diff
@@ -0,0 +1,181 @@
+from mmcv.device.ipu import IS_IPU
+
+if IS_IPU:
+    from mmcv.device.ipu import cast_to_options
+    from mmcv.device.ipu.utils import model_sharding
```
```suggestion
    from mmcv.device import (cast_to_options, model_sharding, build_from_cfg_with_wrapper)
```

HumberMe replied:
model_sharding, build_from_cfg_with_wrapper only used in ipu module, so I didn't put them in mmcv/device/ipu/__init__.py

zhouzaida replied:
got it

zhouzaida commented on tests/test_device/test_ipu/test_ipu_utils.py:
```diff
@@ -0,0 +1,181 @@
+
+
+@skip_no_ipu
+def test_parse_ipu_options():
```
```suggestion
def test_cast_to_options():
```

zhouzaida commented on tests/test_device/test_ipu/test_ipu_utils.py:
```diff
@@ -0,0 +1,181 @@
+        ),
+        eval_cfg=dict(deviceIterations=1, ),
+    )
+    cast_to_options(copy.deepcopy(options_cfg))
```
We need to check whether the output of `cast_to_options` works as expected.

zhouzaida commented on mmcv/device/ipu/ipu_wrapper.py:
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+
+
+def is_ipu():
+    try:
+        import poptorch  # noqa: E261, F401
+        IS_IPU = True
+    except ImportError:
+        IS_IPU = False
+    return IS_IPU
+
+
+IS_IPU = is_ipu()
```
```suggestion
def is_ipu():
    try:
        import poptorch  # noqa: E261, F401
    except ImportError:
        return False
    return True


IS_IPU = is_ipu()
```

Commit:
adjust tests

ZwwWayne commented on mmcv/device/ipu/utils.py:
```diff
@@ -0,0 +1,244 @@
+        f'recomputed nodes: {module_names} are not contained in the model'
+
+
+def compare_tensor(featA, featB, rtol=1e-3, atol=1e-5):
```
suggest renaming it as compare_ndarray, since the data is actually np.ndarray rather than torch.tensor.

HuDi2018 committed:
compare_tensor to compare_ndarray
